indra/llcommon/llcallbacklist.h

```diff
@@ -27,53 +27,248 @@
+inline
+LLCallbackList::handle_t doOnIdleOneTime(nullary_func_t callable)
+{
+    return gIdleCallbacks.doOnIdleOneTime(callable);
+}
 
 // Repeatedly call a callable in idle loop until it returns true.
-void doOnIdleRepeating(bool_func_t callable);
+inline
+LLCallbackList::handle_t doOnIdleRepeating(bool_func_t callable)
+{
+    return gIdleCallbacks.doOnIdleRepeating(callable);
+}
+
+/*****************************************************************************
+*   LLLater: callbacks at some future time
+*****************************************************************************/
+class LLLater: public LLSingleton<LLLater>
+{
+    LLSINGLETON(LLLater);
+
+    using token_t = U32;
+
+    // Define a struct for our priority queue entries, instead of using
+    // a tuple, because we need to define the comparison operator.
+    struct func_at
+    {
+        nullary_func_t mFunc;
+        token_t mToken;
+        LLDate::timestamp mTime;
+
+        func_at(const nullary_func_t& func, token_t token, LLDate::timestamp tm):
+            mFunc(func),
+            mToken(token),
+            mTime(tm)
+        {}
+
+        friend bool operator<(const func_at& lhs, const func_at& rhs)
+        {
+            // use greater-than because we want fibonacci_heap to select the
+            // EARLIEST time as the top()
+            return lhs.mTime > rhs.mTime;
+        }
+    };
+
+    // Accept default stable<false>: when two funcs have the same timestamp,
+    // we don't care in what order they're called.
+    // Specify constant_time_size<false>: we don't need to optimize the size()
+    // method, iow we don't need to store and maintain a count of entries.
+    typedef boost::heap::fibonacci_heap<func_at, boost::heap::constant_time_size<false>>
+        queue_t;
+
+public:
+    // If tasks that come ready during a given tick() take longer than this,
+    // defer any subsequent ready tasks to a future tick() call.
+    static constexpr F32 TIMESLICE{ 0.005f };
+
+    class handle_t
+    {
+    private:
+        friend class LLLater;
+        token_t token;
+    public:
+        handle_t(token_t token=0): token(token) {}
+        bool operator==(const handle_t& rhs) const { return this->token == rhs.token; }
+        explicit operator bool() const { return bool(token); }
+        bool operator!() const { return ! bool(*this); }
+    };
 
+    // Call a given callable once at specified timestamp.
+    handle_t doAtTime(nullary_func_t callable, LLDate::timestamp time);
+
+    // Call a given callable once after specified interval.
+    handle_t doAfterInterval(nullary_func_t callable, F32 seconds);
+
+    // Call a given callable every specified number of seconds, until it returns true.
+    handle_t doPeriodically(bool_func_t callable, F32 seconds);
+
+    // test whether specified handle is still live
+    bool isRunning(handle_t timer) const;
+    // check remaining time
+    F32 getRemaining(handle_t timer) const;
```

Comment:
Here is a good example why I am musing about a specific `seconds` type. `getRemaning` what? Also it is not obvious what this means:
- remaining units till this handle executes?
- remaining units till this handle finishes execution?
- remaining execution counts?

Reply:
The function I'm trying to emulate here is [`F32SecondsImplicit getRemainingTimeF32()`](https://github.com/secondlife/viewer/blob/main/indra/llcommon/lltimer.h#L100). I just think that's an awkwardly long name. I grant that `getRemainingSecondsF32()` would be even clearer, but it's even longer.

Again, the name is based on the idea that a `handle_t` references a timer, and it returns the time until that timer is supposed to fire. What the callback chooses to do at that point is up to the callback.

The second bullet isn't a reasonable interpretation. Even with reflection this class can't know how long until the callback finishes running: you run smack into The Halting Problem.

Since a `doPeriodically()` callback's `bool` return determines whether it will be called again in future, similar remarks apply to the third bullet.

Reply:
In times for auotcompletion everywhere and even AI suggesting us code (to then fix the mistakes the AI made) I'd argue a longer name to make intent clear is worth it